SimPEG/regularization/coupling.py

```diff
@@ -0,0 +1,744 @@
+        Returns either an array of dimensions [k*M,1] (v is not None), or
+        sparse matrix of dimensions [k*M, k*M] (v is None).
+        k: number of models we are inverting for.
+        M: number of cells in each model.
+
+        '''
+        raise NotImplementedError(
+            "The method _deriv2 has not been implemented for {}".format(
+                self.__class__.__name__
+            )
+        )
+
+    def __call__(self):
+        ''' Returns the computed value of the coupling term. '''
+        raise NotImplementedError(
+            "The method __call__ has not been implemented for {}".format(
+                self.__class__.__name__
+            )
+        )
+
+
+###############################################################################
+#                                                                             #
+#                                Cross-Gradient                               #
+#                                                                             #
+###############################################################################
+
+
+class CrossGradient(BaseCoupling):
+    '''
+    The cross-gradient constraint for joint inversions.
+
+    ..math::
+        \phi_c(\mathbf{m_1},\mathbf{m_2}) = \lambda \sum_{i=1}^{M} \|
+        \nabla \mathbf{m_1}_i \times \nabla \mathbf{m_2}_i \|^2
+
+    All methods assume that we are working with two models only.
+
+    '''
+    def __init__(self, mesh, indActive, mapping, **kwargs):
+        self.as_super = super(CrossGradient, self)
+        self.as_super.__init__(mesh, indActive, mapping, **kwargs)
+        self.map1, self.map2 = mapping.maps # Assume a map has been passed for each model.
+
+        assert mesh.dim in (2,3), 'Cross-Gradient is only defined for 2D or 3D'
+
+    def models(self, ind_models):
+        '''
+        Method to pass models to CrossGradient object and ensures models are
+        compatible for further use. Checks that models are of same size.
+
+        :param container of numpy.ndarray ind_models: [model1, model2,...]
+
+        rtype: list of numpy.ndarray models: [model1, model2,...]
+        return: models
+
+        '''
+        models = []
+        n = len(ind_models)
+        for i in range(n):
+            # check that the models are either a list, tuple, or np.ndarray
+            assert isinstance(ind_models[i], (list,tuple,np.ndarray))
+            if isinstance(ind_models[i], (list,tuple)):
+                ind_models[i] = np.array(ind_models[i])
+
+        # check if models are of same size
+        it = iter(ind_models)
+        length = len(next(it))
+        if not all(len(l)==length for l in it):
+            raise ValueError('not all models are of the same size!')
+
+        for i in range(n):
+            models.append(ind_models[i])
+
+        return models
+
+    def calculate_gradient(self, model):
+        '''
+        Calculate the spatial gradients of the model using central difference.
```

Comment:
It is true. I have had a discussion about it with someone else on the discourse (https://simpeg.discourse.group/t/how-to-define-a-difference-operator-mapping-from-cell-center-to-cell-center/149/4)

Basically, a checkboard is in the null-space of the central difference operator.

Reply:
Hi, Joe, thanks for your comments that are extremely helpful. 

Based on the Jae and my work, the central difference operator works fine for cross-gradient here. But I did see some checkerboard like patterns (figure below) in joint total variation (another joint inversion method I worked on), where I use central difference operator as well. Maybe what you said is the reason. 
![plt_model_rec_dens](https://user-images.githubusercontent.com/56940558/101843149-34201c80-3b0f-11eb-8720-314e2ef83c76.png)

Reply:
I think it would be helpful to get the forward gradient operator in discretize. It may help solve some issues. 

If JTV method can works well using forward operator, that would be great! As I already temporally give up this method for this issue.

Reply:
> It is true. I have had a discussion about it with someone else on the discourse (https://simpeg.discourse.group/t/how-to-define-a-difference-operator-mapping-from-cell-center-to-cell-center/149/4)
> 
> Basically, a checkboard is in the null-space of the central difference operator.

Really? This is really helpful but new to me.

Reply:
One approach to consider here is using the use of the vector identity for the cross product as suggested by [(Haber and Gazit, 2013)](https://idp.springer.com/authorize/casa?redirect_uri=https://link.springer.com/content/pdf/10.1007/s10712-013-9232-4.pdf&casa_token=ezeDmAwQuhsAAAAA:hO3GzSpdJDIRxtrknNkW6QTOShEvOrc-oE3r6ceV6cA1ScC0xzZoYk9JjyO5nglF0pqjpJkChn4KBa7Y)

The relevant discussion is around equation 11:

![image](https://user-images.githubusercontent.com/6361812/101861650-c6cfb400-3b2d-11eb-82b5-0c271a3132a0.png)

Reply:
Thanks for comment! @lheagy . It's a really good paper and we did follow the method mentioned in Dr. Haber's paper. Please check the equation under "def__call__" function.

Reply:
Just to add to the discussion. I can confirm we followed the approach in Dr. Haber's paper. And also, we didn't encounter any problems regarding checkerboard-like artifacts (like those shown by Xiaolong) that couldn't be solved by adjusting the weight of the cross-gradient term. 
But I can see where there might be problems when facing a checkboard-like model where periodic values (`m = [-1, 1,-1, 1, [...] ],`) might lead to zero gradients. Indeed, the central difference method will give zero gradients in these instances.
The forward difference operators could work in such cases, but this will likely require reformulating the cross-gradient computations in the code, since the dimensions of the gradients will be different in the x-, y-, and z-directions (This is the same issue in https://simpeg.discourse.group/t/how-to-define-a-difference-operator-mapping-from-cell-center-to-cell-center/149/4). I believe there will be issues in places like lines 151, 159, 397, ...
Perhaps the third strategy @xiaolongw1223 mentions in line 322 might work with forward difference operators, but I'm not sure of this.

Reply:
I don't think it will change anything in the code to switch to a forward difference operator, as this code is actually fairly generalized to the gradient matrix. We would just need an operator that does that forward gradient, where the result lives on the cell.

Reply:
Agreed. If the results live on the cell centers then I see no problem.